Sources/Foundation/JSONSerialization.swift

```diff
@@ -549,523 +615,913 @@ private struct JSONWriter {
 }
 
 //MARK: - JSONDeserializer
-private struct JSONReader {
-
-    static let whitespaceASCII: [UInt8] = [
-        0x09, // Horizontal tab
-        0x0A, // Line feed or New line
-        0x0D, // Carriage return
-        0x20, // Space
-    ]
-
-    struct Structure {
-        static let BeginArray: UInt8     = 0x5B // [
-        static let EndArray: UInt8       = 0x5D // ]
-        static let BeginObject: UInt8    = 0x7B // {
-        static let EndObject: UInt8      = 0x7D // }
-        static let NameSeparator: UInt8  = 0x3A // :
-        static let ValueSeparator: UInt8 = 0x2C // ,
-        static let QuotationMark: UInt8  = 0x22 // "
-        static let Escape: UInt8         = 0x5C // \
+struct JSONParser {
+    init() {}
+
+    @inlinable
+    func parse<Bytes: Collection>(bytes: Bytes) throws
+        -> JSONValue where Bytes.Element == UInt8
+    {
+        var impl = JSONParserImpl(bytes: bytes)
+        return try impl.parse()
     }
+}
 
-    typealias Index = Int
-    typealias IndexDistance = Int
-
-    struct UnicodeSource {
-        let buffer: UnsafeBufferPointer<UInt8>
-        let encoding: String.Encoding
-        let step: Int
-
-        init(buffer: UnsafeBufferPointer<UInt8>, encoding: String.Encoding) {
-            self.buffer = buffer
-            self.encoding = encoding
-
-            self.step = {
-                switch encoding {
-                case .utf8:
-                    return 1
-                case .utf16BigEndian, .utf16LittleEndian:
-                    return 2
-                case .utf32BigEndian, .utf32LittleEndian:
-                    return 4
-                default:
-                    return 1
-                }
-            }()
+@usableFromInline struct JSONParserImpl {
```

Comment:
Did each of these `@usableFromInline` show a benefit in benchmarks? If not I would avoid using them otherwise.

However one thing it is worth doing is tightening up the visibility of these types (`private`, `fileprivate`) since the compiler will use that information for inlining as well. If the compiler knows it doesn't have to make the type/method/ivar visible outside of this file then that gives it more hints for better code generation.

Reply:
Fixed.